Rollbar/RollbarQueueController.cs

```diff
@@ -0,0 +1,298 @@
+        private RollbarQueueController()
+        {
+            this._rollbarCommThread = new Thread(this.KeepProcessingAllQueues);
+            this._rollbarCommThread.IsBackground = true;
+            this._rollbarCommThread.Name = "Rollbar Communication Thread";
+
+            this._rollbarCommThread.Start();
+        }
+
+        private sealed class NestedSingleInstance
+        {
+            static NestedSingleInstance()
+            {
+            }
+
+            internal static readonly RollbarQueueController Instance = 
+                new RollbarQueueController();
+        }
+
+        #endregion singleton implementation
+
+        /// <summary>
+        /// Occurs after a Rollbar internal event happens.
+        /// </summary>
+        public event EventHandler<RollbarEventArgs> InternalEvent;
+
+        /// <summary>
+        /// Registers the specified queue.
+        /// </summary>
+        /// <param name="queue">The queue.</param>
+        internal void Register(PayloadQueue queue)
+        {
+            lock (this._syncLock)
+            {
+                Assumption.AssertTrue(!this._allQueues.Contains(queue), nameof(queue));
+
+                this._allQueues.Add(queue);
+                this.IndexByToken(queue);
+                queue.Logger.Config.Reconfigured += Config_Reconfigured;
+                Debug.WriteLine(this.GetType().Name + ": Registered a queue. Total queues count: " + this._allQueues.Count + ".");
+            }
+        }
+
+        /// <summary>
+        /// Unregisters the specified queue.
+        /// </summary>
+        /// <param name="queue">The queue.</param>
+        internal void Unregister(PayloadQueue queue)
+        {
+            lock (this._syncLock)
+            {
+                Assumption.AssertTrue(!queue.Logger.IsSingleton, nameof(queue.Logger.IsSingleton));
+                Assumption.AssertTrue(this._allQueues.Contains(queue), nameof(queue));
+
+                this.DropIndexByToken(queue);
+                this._allQueues.Remove(queue);
+                queue.Logger.Config.Reconfigured -= Config_Reconfigured;
+                Debug.WriteLine(this.GetType().Name + ": Unregistered a queue. Total queues count: " + this._allQueues.Count + ".");
+            }
+        }
+
+        private readonly object _syncLock = new object();
+
+        private readonly Thread _rollbarCommThread = null;
+
+        private readonly HashSet<PayloadQueue> _allQueues =
+            new HashSet<PayloadQueue>();
+
+        private readonly Dictionary<string, AccessTokenQueuesMetadata> _queuesByAccessToken = 
+            new Dictionary<string, AccessTokenQueuesMetadata>();
+
+        private void KeepProcessingAllQueues()
+        {
+            TimeSpan sleepInterval = TimeSpan.FromMilliseconds(250);
+
+            while(true)
+            {
+                try
+                {
+                    lock(this._syncLock)
+                    {
+                        ProcessAllQueuesOnce();
```

Comment:
Agree, that would be right optimization to consider. But, first I want to make sure that the "big design" is acceptable before spending too much effort on optimization. Also, remember, that it is not about just pulling items from a queue but from a dictionary of a queue lists. Depending how and how often users choose to create and use instances of IRollbar, the dictionary and its lists could be modified from different threads. Also, I recall we were talking about making the queues fixed-depth-queues. That would imply that any enqueue operation may potentially cause dequeue op. All these aspects will have to be carefully addressed during the optimization iteration(s).

Reply:
I usually wouldn't consider it an optimization because of the potential for a deadlock associated with calling out while holding a lock, but as the code is written I don't see a deadlock and as long as it is on the radar to fix later then I'm okay with this.

Reply:
Right now, locking (using the _syncLock instance) resolves concurrency among 4 major types of operations that may affect the queues controller: 1. de/registration of a token; 2. de/registration of a queue; 3. queuing an payload;  4. processing queue (de-queuing payloads) that are in general all could be preformed from different threads. It is basically mutually excludes all but one out all the threads. All the operations except ProcessAllQueuesOnce() are very quick and should scale out well. The queues processing operation depends on total number of tokens, queues for each token and number Rollbar API communication retries (if any). If we introduce new locks dedicated to each of the following just token de/registration, queue de/registration per token, and payload de/queuing per queue, we would not have to lock for "so long" during the entire queue processing but will be able to unlock it more often and lock it for shorter sub-steps of queues processing. Hence, the optimization is the word here. However, introduction of such a granular fine tuned locking using multiple dedicated syncLocks makes things more complex and opens more opportunities to introduction of unintended deadlocks. That is exactly what optimizations do: make things more performant but also more complex and less straightforward. :)